core/indexer/component.go

```diff
@@ -0,0 +1,174 @@
+package indexer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"go.uber.org/dig"
+
+	"github.com/gohornet/inx-indexer/pkg/daemon"
+	"github.com/gohornet/inx-indexer/pkg/indexer"
+	"github.com/gohornet/inx-indexer/pkg/nodebridge"
+	"github.com/gohornet/inx-indexer/pkg/server"
+	"github.com/iotaledger/hive.go/app"
+	"github.com/iotaledger/hive.go/app/core/shutdown"
+	inx "github.com/iotaledger/inx/go"
+)
+
+const (
+	APIRoute = "indexer/v1"
+)
+
+func init() {
+	CoreComponent = &app.CoreComponent{
+		Component: &app.Component{
+			Name:     "Indexer",
+			DepsFunc: func(cDeps dependencies) { deps = cDeps },
+			Params:   params,
+			Provide:  provide,
+			Run:      run,
+		},
+	}
+}
+
+type dependencies struct {
+	dig.In
+	NodeBridge      *nodebridge.NodeBridge
+	Indexer         *indexer.Indexer
+	ShutdownHandler *shutdown.ShutdownHandler
+}
+
+var (
+	CoreComponent *app.CoreComponent
+	deps          dependencies
+)
+
+func provide(c *dig.Container) error {
+
+	if err := c.Provide(func() (*indexer.Indexer, error) {
+		CoreComponent.LogInfo("Setting up database...")
+		return indexer.NewIndexer(ParamsIndexer.Database.Path)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func run() error {
+	// create a background worker that handles the indexer events
+	CoreComponent.Daemon().BackgroundWorker("Indexer", func(ctx context.Context) {
+		CoreComponent.LogInfo("Starting Indexer")
+		defer deps.Indexer.CloseDatabase()
+
+		fillIndexer := false
+
+		// Checking initial indexer state
+		ledgerIndex, err := deps.Indexer.LedgerIndex()
+		if err != nil {
+			if !errors.Is(err, indexer.ErrNotFound) {
+				CoreComponent.LogPanicf("Reading ledger index from Indexer failed! Error: %s", err)
+				return
+			}
+			CoreComponent.LogInfo("Indexer is empty, so import initial ledger...")
+			fillIndexer = true
+		}
+
+		if !fillIndexer && deps.NodeBridge.LedgerPruningIndex() > ledgerIndex {
+			CoreComponent.LogInfo("> Node has an newer pruning index than our current ledgerIndex")
+			CoreComponent.LogInfo("Re-import initial ledger...")
+
+			if err := deps.Indexer.Clear(); err != nil {
+				CoreComponent.LogPanicf("Clearing Indexer failed! Error: %s", err)
+				return
+			}
+			fillIndexer = true
+		}
+
+		if fillIndexer {
+			// Indexer is empty, so import initial ledger state from the node
+			if err := deps.NodeBridge.FillIndexer(ctx, deps.Indexer); err != nil {
+				CoreComponent.LogPanicf("Filling Indexer failed! Error: %s", err)
+				return
+			}
+			CoreComponent.LogInfof("Imported initial ledger at index %d", ledgerIndex)
```

Comment:
this ledgerIndex should probably be re-read from `deps.Indexer` after it was filled